tuftool/src/source.rs

```diff
@@ -8,169 +8,54 @@
-
-    #[cfg(any(feature = "rusoto-native-tls", feature = "rusoto-rustls"))]
-    fn write_with_ssm_key(
-        value: &str,
-        key_id_hex: &str,
-        profile: &Option<String>,
-        parameter_name: &str,
-        key_id: &Option<String>,
-    ) -> Result<()> {
-        use crate::deref::OptionDeref;
-        use rusoto_ssm::Ssm;
-
-        let ssm_client = crate::ssm::build_client(profile.deref_shim())?;
-        let fut = ssm_client.put_parameter(rusoto_ssm::PutParameterRequest {
-            name: parameter_name.to_owned(),
-            description: Some(key_id_hex.to_owned()),
-            key_id: key_id.as_ref().cloned(),
-            overwrite: Some(true),
-            type_: "SecureString".to_owned(),
-            value: value.to_owned(),
-            ..rusoto_ssm::PutParameterRequest::default()
-        });
-        tokio::runtime::Runtime::new()
-            .unwrap()
-            .block_on(fut)
-            .context(error::SsmPutParameter {
-                profile: profile.clone(),
-                parameter_name,
-            })?;
-        Ok(())
-    }
-}
-
-impl FromStr for KeySource {
-    type Err = Error;
-
-    #[allow(clippy::find_map)]
-    fn from_str(s: &str) -> Result<Self> {
-        let pwd_url = Url::from_directory_path(std::env::current_dir().context(error::CurrentDir)?)
-            .expect("expected current directory to be absolute");
-        let url = Url::options()
-            .base_url(Some(&pwd_url))
-            .parse(s)
-            .context(error::UrlParse { url: s })?;
-
-        match url.scheme() {
-            "file" => Ok(KeySource::Local(PathBuf::from(url.path()))),
-            #[cfg(any(feature = "rusoto-native-tls", feature = "rusoto-rustls"))]
-            "aws-ssm" => Ok(KeySource::Ssm {
-                profile: url.host_str().and_then(|s| {
-                    if s.is_empty() {
-                        None
-                    } else {
-                        Some(s.to_owned())
-                    }
-                }),
-                parameter_name: url.path().to_owned(),
-                key_id: url
-                    .query_pairs()
-                    .find(|(k, _)| k == "kms-key-id")
-                    .map(|(_, v)| v.into_owned()),
+/// Parses a user-specified source of signing keys.
+/// Currently supported key sources are local files and AWS SSM.
+/// Sources are passed to `tuftool` as arguments in string format:
+/// "file:///" or "aws-ssm:///"
+///
+/// Users are welcome to add their own sources of keys by implementing
+/// the `KeySource` trait in the `tough` library. A user can then add
+/// to this parser to support them in `tuftool`.
+pub(crate) fn parse_key_source(input: &str) -> Result<Box<dyn KeySource>> {
+    let pwd_url = Url::from_directory_path(std::env::current_dir().context(error::CurrentDir)?)
+        .expect("expected current directory to be absolute");
+    let url = Url::options()
+        .base_url(Some(&pwd_url))
+        .parse(input)
```

Comment:
Why use `options`, `base_url`, and `parse`, if we're not changing the options?  Wouldn't `Url::parse` do the same?  Is it related to the (unrecognized?) `aws-ssm` scheme?  In any case, can we comment why we're setting the base URL to the current directory?

Reply:
The base URL is required in the case of files, otherwise the `URL` lib complains about a "relative URL without a base".  This is actually mentioned in the doc comment for this file, but I'll move it down so it's closer to the applicable code.